crates/fuel-core/src/graphql_api/da_compression.rs

```diff
@@ -0,0 +1,190 @@
+    graphql_api::storage::da_compression::*,
+};
+use fuel_core_compression::{
+    compress::compress,
+    ports::{
+        EvictorDb,
+        TemporalRegistry,
+        UtxoIdToPointer,
+    },
+};
+use fuel_core_storage::{
+    not_found,
+    StorageAsMut,
+    StorageAsRef,
+};
+use fuel_core_types::{
+    blockchain::block::Block,
+    fuel_tx::{
+        input::PredicateCode,
+        Address,
+        AssetId,
+        ContractId,
+        ScriptCode,
+    },
+    services::executor::Event,
+};
+use futures::FutureExt;
+use sha2::{
+    Digest,
+    Sha256,
+};
+
+/// Performs DA compression for a block and stores it in the database.
+pub fn da_compress_block<T>(
+    block: &Block,
+    block_events: &[Event],
+    db_tx: &mut T,
+) -> anyhow::Result<()>
+where
+    T: OffChainDatabaseTransaction,
+{
+    let compressed = compress(
+        CompressTx {
+            db_tx,
+            block_events,
+        },
+        block,
+    )
+    .now_or_never()
+    .expect("The current implementation resolved all futures instantly")?;
+
+    db_tx
+        .storage_as_mut::<DaCompressedBlocks>()
+        .insert(&block.header().consensus().height, &compressed)?;
+
+    Ok(())
+}
+
+struct CompressTx<'a, Tx> {
+    db_tx: &'a mut Tx,
+    block_events: &'a [Event],
+}
+
+macro_rules! impl_temporal_registry {
+    ($type:ty, $index_value_fn:expr) => { paste::paste! {
+        impl<'a, Tx> TemporalRegistry<$type> for CompressTx<'a, Tx>
+        where
+            Tx: OffChainDatabaseTransaction,
+        {
+            fn read_registry(
+                &self,
+                key: &fuel_core_types::fuel_compression::RegistryKey,
+            ) -> anyhow::Result<$type> {
+                Ok(self
+                    .db_tx
+                    .storage_as_ref::<[< DaCompressionTemporalRegistry $type >]>()
+                    .get(key)?
+                    .ok_or(not_found!([< DaCompressionTemporalRegistry $type>]))?
+                    .into_owned())
+            }
+
+            fn write_registry(
+                &mut self,
+                key: &fuel_core_types::fuel_compression::RegistryKey,
+                value: &$type,
+            ) -> anyhow::Result<()> {
+                // Write the actual value
+                self.db_tx
+                    .storage_as_mut::<[< DaCompressionTemporalRegistry $type >]>()
+                    .insert(key, value)?;
+
+                let value_in_index: [u8; 32] = ($index_value_fn)(value);
+
+                // Remove the overwritten value from index, if any
+                self.db_tx
+                    .storage_as_mut::<[< DaCompressionTemporalRegistryIndex $type >]>()
+                    .remove(&value_in_index)?;
+
+                // Add the new value to the index
+                self.db_tx
+                    .storage_as_mut::<[< DaCompressionTemporalRegistryIndex $type >]>()
+                    .insert(&value_in_index, key)?;
```

Comment:
Why do we need to remove it first and insert after? Insert will replace the value automaticly

Reply:
This was broken in some refactor. Fixed in ee3e844626.